### PR TITLE
AWS common: region handling cleanup

### PR DESCRIPTION
**Pull request by nbaws**
Commit Message: AWS common: region handling cleanup
Additional Description: Refactor region handling code as per https://github.com/envoyproxy/envoy/issues/31676
Risk Level: Low
Testing: Unit
Docs Changes: Yes, new region documentation describes region string retrieval process
Release Notes: Changelog
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] 31676
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

This PR uplifts some of the AWS common code to ensure regions are read consistently from environment and config/credentials file across extensions.

The following changes were made to accommodate this:

- Move environment variable/default code from credential_provider_impl.cc to utility.cc
- Move config/credentials file parsing code from credential_provider_impl.cc to utility.cc and make this generic to support both credentials and region use cases
- Create support functions in utility.cc to resolve a profile string and credentials/config file paths
- Update region_provider_impl to provide a region chain, that retrieves region string from environment/config/credentials file
- Update aws_request_signer extension to use the region chain if region is not explicit in envoy config, and make the regional optional in protobuf 
- Update grpc_credentials/aws_iam extension to use the region chain if region is not explicit in envoy config

### CODE REVIEW

**Comment by nbaws** [comment]
@suniltheta

File: changelogs/current.yaml [hunk 1]
```diff
@@ -45,5 +45,9 @@ new_features:
   change: |
     added support for :ref:`%UPSTREAM_CONNECTION_ID% <config_access_log_format_upstream_connection_id>` for the upstream connection
     identifier.
+- area: aws
```
Comment (suniltheta) [inline changelogs/current.yaml:48]: do you want to mention this under minor behavior change? Also can mention that region field is made optional and explicitly configured value will be preferred.

File: source/extensions/common/aws/region_provider_impl.cc [hunk 2]
```diff
@@ -1,23 +1,95 @@
...
 namespace Extensions {
 namespace Common {
 namespace Aws {
 
 static const char AWS_REGION[] = "AWS_REGION";
+static const char AWS_DEFAULT_REGION[] = "AWS_DEFAULT_REGION";
+constexpr char REGION[] = "REGION";
+
+absl::optional<std::string> EnvironmentRegionProvider::getRegion() {
+  ENVOY_LOG_MISC(debug, "called EnvironmentRegionProvider::getRegion");
```
Comment (suniltheta) [inline source/extensions/common/aws/region_provider_impl.cc:15]: Can use `__func__` here if these are useful while development. Also prefer if they are just trace level instead of debug. Same for logs on other functions.

You can also consider logging in debug level once it is successfully read from particular region provider.

File: source/extensions/filters/http/aws_request_signing/config.cc [hunk 3]
```diff
@@ -51,27 +52,45 @@ AwsRequestSigningFilterFactory::createFilterFactoryFromProtoTyped(
...
 
+  std::string region;
+  region = config.region();
+
+  if (region.empty()) {
+    auto region_provider = std::make_shared<Extensions::Common::Aws::RegionProviderChain>();
+    absl::optional<std::string> regionopt = region_provider->getRegion();
+    if (!regionopt.has_value()) {
+      throw EnvoyException("Region string cannot be retrieved from configuration, environment or "
+                           "profile/config files.");
```
Comment (suniltheta) [inline source/extensions/filters/http/aws_request_signing/config.cc:63]: Do you want to consider changing the exception message to something like below ?
`"AWS region is not set in xDS configuration and failed to retrieve from environment variable or AWS profile/config files."`

File: source/extensions/filters/http/aws_request_signing/config.cc [hunk 4]
```diff
@@ -51,27 +52,45 @@ AwsRequestSigningFilterFactory::createFilterFactoryFromProtoTyped(
     const AwsRequestSigningProtoConfig& config, const std::string& stats_prefix, DualInfo dual_info,
     Server::Configuration::ServerFactoryContext& server_context) {
 
+  std::string region;
+  region = config.region();
+
+  if (region.empty()) {
+    auto region_provider = std::make_shared<Extensions::Common::Aws::RegionProviderChain>();
+    absl::optional<std::string> regionopt = region_provider->getRegion();
```
Comment (suniltheta) [inline source/extensions/filters/http/aws_request_signing/config.cc:60]: nit: `regionOpt`

nit: `auto` instead of `absl::optional<std::string>` is possible?

File: source/extensions/filters/http/aws_request_signing/config.cc [hunk 5]
```diff
@@ -51,27 +52,45 @@ AwsRequestSigningFilterFactory::createFilterFactoryFromProtoTyped(
...
       config.match_excluded_headers().begin(), config.match_excluded_headers().end());
 
   std::unique_ptr<Extensions::Common::Aws::Signer> signer;
 
   if (getSigningAlgorithm(config) == SigningAlgorithm::SIGV4A) {
+    if (config.region().empty()) {
+      throw EnvoyException("Region parameter does not contain a SigV4A region set.");
+    }
+    // We use config.region here as environment or file store is not a valid region location for
+    // AWS_SIGV4A
```
Comment (suniltheta) [inline source/extensions/filters/http/aws_request_signing/config.cc:82]: why region set cannot be ready from Env or Config file?
Can you explicitly call this limitation from Envoy docs associated with Sigv4A signing? Also mention the same on proto that region is not option for SIGV4A signing.

File: source/extensions/filters/http/aws_request_signing/config.cc (hunk 3 above)
Comment (suniltheta) [reply source/extensions/filters/http/aws_request_signing/config.cc:63]: Similar exception message inside perRoute function

File: source/extensions/filters/http/aws_request_signing/config.cc (hunk 5 above)
Comment (nbaws) [reply source/extensions/filters/http/aws_request_signing/config.cc:82]: there is currently no defined sdk behaviour for region sets. so the safest option is not to treat an environment/config/credentials retrieved region as valid.
will update proto to make this more clear

File: source/extensions/filters/http/aws_request_signing/config.cc (hunk 4 above)
Comment (nbaws) [reply source/extensions/filters/http/aws_request_signing/config.cc]: done

File: source/extensions/filters/http/aws_request_signing/config.cc (hunk 3 above)
Comment (nbaws) [reply source/extensions/filters/http/aws_request_signing/config.cc]: done

File: source/extensions/common/aws/region_provider_impl.cc (hunk 2 above)
Comment (nbaws) [reply source/extensions/common/aws/region_provider_impl.cc]: done

File: changelogs/current.yaml (hunk 1 above)
Comment (nbaws) [reply changelogs/current.yaml]: done

File: source/extensions/filters/http/aws_request_signing/config.cc (hunk 5 above)
Comment (nbaws) [reply source/extensions/filters/http/aws_request_signing/config.cc:82]: i've also clarified this more explicitly in all doc references.

File: api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto [hunk 6]
```diff
@@ -39,6 +39,10 @@ message AwsRequestSigning {
   // Example: s3
   string service_name = 1 [(validate.rules).string = {min_len: 1}];
 
+  // Optional region string. If region is not provided, the region will be retrieved from the environment
+  // or AWS configuration files. The region string is mandatory if signing_algorithm is set to ``AWS_SIGV4A`` and will
+  // be checked at runtime.
```
Comment (wbpcode) [inline api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto:44]: What will be preferred between ENV and configuration file?

Comment (nbaws) [reply api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto:44]: ok - there will be some slight behaviour changes as I've received confirmation from the service team as to how this can be configured environment/config-wise. i also documented the behaviour in https://github.com/envoyproxy/envoy/pull/32070/files#diff-f75c1708c2e1316d0d66a9b1af39bf0ca022c116ea97f6403826372be3699a53 , however i will xref it here to make it very clear.

File: test/per_file_coverage.sh
Comment (nbaws) [inline test/per_file_coverage.sh:1]: Coverage lowered - this change is unrelated to the flagged coverage misses.

File: api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto (hunk 6 above)
Comment (wbpcode) [reply api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto]: oh, I see. Thanks, the docs is very clear. Then I think you can add a link to the docs in the proto field comments.

File: test/per_file_coverage.sh
Comment (RyanTheOptimist) [reply test/per_file_coverage.sh:1]: I'm not sure that I understand. It likes this PR changes code in `source/extensions/grpc_credentials/aws_iam` and the coverage is being lowered for the same directory: "source/extensions/grpc_credentials/aws_iam:95.4"

Can we add sufficient tests to avoid lowering the coverage limit, please?

Comment (nbaws) [reply test/per_file_coverage.sh:1]: What I meant was that the code that is untested is not the changed code. The untested lines were also untested before my change, just the % changed due to the PR.

That being said I've added a test case that I hope should cover this for any future changes.

File: api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto (hunk 6 above)
Comment (nbaws) [reply api/envoy/extensions/filters/http/aws_request_signing/v3/aws_request_signing.proto]: should be in there now :)

File: test/per_file_coverage.sh
Comment (RyanTheOptimist) [reply test/per_file_coverage.sh:1]: Great, thanks! Since coverage is no longer being updated, I'll remove myself as coverage-shephards assignee.

File: source/extensions/filters/http/aws_request_signing/config.cc (hunk 5 above)
Comment (nbaws) [reply source/extensions/filters/http/aws_request_signing/config.cc]: > there is currently no defined sdk behaviour for region sets. so the safest option is not to treat an environment/config/credentials retrieved region as valid. will update proto to make this more clear

I've now had confirmation that there is in fact defined behaviour - which is to use `AWS_SIGV4A_SIGNING_REGION_SET` or profile/credentials string `sigv4a_signing_region_set` . This is now the implementation in this PR, rather than exception if region not specific for SigV4A

**Comment by alyssawilk** [comment]
throwing over to @mattklein123 for aws common review

**Review by suniltheta — APPROVED** [review]
Thanks for accommodating SIGV4A_SIGNING_REGION_SET as well

**Review by mattklein123 — APPROVED** [review]

**Comment by mattklein123** [comment]
Oops sorry can you merge main.

/wait

**Comment by nbaws** [comment]
> Oops sorry can you merge main.
> 
> /wait

Done

**Review by mattklein123 — APPROVED** [review]
